server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -293,6 +294,34 @@ static void loadExtensions(Collection<LoadedPlugin> plugins) {
         }
     }
 
+    // package private for testing
+    <T> Iterator<T> providersIterator(Class<T> service, ClassLoader classLoader) {
+        return ServiceLoader.load(service, classLoader).iterator();
+    }
+
+    /**
+     * SPI convenience method that uses the {@link ServiceLoader} JDK class to load various SPI providers
+     * from plugins/modules.
+     *
+     * <p>
+     * For example:
+     *
+     * var pluginHandlers = pluginsService.loadServiceProviders(OperatorHandlerProvider.class);
+     * </p>
+     * @param service A templated service class to look for providers in plugins
+     * @return an immutable {@link List} of discovered providers in the plugins/modules
+     */
+    public <T> List<? extends T> loadServiceProviders(Class<T> service) {
+        List<T> result = new ArrayList<>();
+        getClass().getModule().addUses(service);
```

Comment:
Why is this necessary? If a type is known in server, then it can be declared explicitly in server's module-info?

Reply:
I thought of it as a convenience, but we can always explicitly declare the use in module-info. I'll remove it.